sendrecv/gst-rust/src/main.rs

```diff
@@ -0,0 +1,480 @@
+    );
+    app_control
+        .ws_sender
+        .send(format!("SESSION {}", app_control.peer_id))
+        .unwrap();
+    return AppState::PeerConnecting;
+}
+
+fn register_with_server(app_control: &Arc<Mutex<AppControl>>) -> AppState {
+    let mut app_control = app_control.lock().unwrap();
+    app_control.app_state = AppState::ServerRegistering;
+    let our_id = rand::thread_rng().gen_range(10, 10_000);
+    println!("Registering id {} with server", our_id);
+    app_control
+        .ws_sender
+        .send(format!("HELLO {}", our_id))
+        .unwrap();
+    return AppState::ServerRegistering;
+}
+
+fn send_sdp_offer(
+    app_control: &Arc<Mutex<AppControl>>,
+    offer: gstreamer_webrtc::WebRTCSessionDescription,
+) {
+    let app_control = app_control.lock().unwrap();
+    if app_control.app_state < AppState::PeerCallNegotiating {
+        // TODO signal and cleanup
+        panic!("Can't send offer, not in call");
+    };
+    let message = json!({
+      "sdp": {
+        "type": "offer",
+        "sdp": offer.get_sdp().as_text().unwrap(),
+      }
+    });
+    app_control.ws_sender.send(message.to_string()).unwrap();
+}
+
+fn on_offer_created(
+    app_control: &Arc<Mutex<AppControl>>,
+    webrtc: gst::Element,
+    promise: &gst::Promise,
+) {
+    assert_eq!(
+        app_control.lock().unwrap().app_state,
+        AppState::PeerCallNegotiating
+    );
+    let reply = promise.get_reply().unwrap();
+
+    let offer = reply
+        .get_value("offer")
+        .unwrap()
+        .get::<gstreamer_webrtc::WebRTCSessionDescription>()
+        .expect("Invalid argument");
+    webrtc
+        .emit("set-local-description", &[&offer, &None::<gst::Promise>])
+        .unwrap();
+
+    send_sdp_offer(app_control, offer)
+}
+
+fn on_negotiation_needed(
+    app_control: &Arc<Mutex<AppControl>>,
+    values: &[glib::Value],
+) -> Option<glib::Value> {
+    app_control.lock().unwrap().app_state = AppState::PeerCallNegotiating;
+    let webrtc = values[0].get::<gst::Element>().expect("Invalid argument");
+    let webrtc_clone = webrtc.clone();
+    let app_control_clone = app_control.clone();
+    let promise = gst::Promise::new_with_change_func(move |promise: &gst::Promise| {
+        on_offer_created(&app_control_clone, webrtc, promise);
+    });
+    let options = gst::Structure::new_empty("options");
+    webrtc_clone
+        .emit("create-offer", &[&options, &promise])
+        .unwrap();
+    None
+}
+
+fn handle_media_stream(
+    pad: &gst::Pad,
+    pipe: &gst::Element,
+    convert_name: &str,
+    sink_name: &str,
+) -> Result<(), Error> {
+    println!(
+        "Trying to handle stream with {} ! {}",
+        convert_name, sink_name,
+    );
+    let q = gst::ElementFactory::make("queue", None).unwrap();
+    let conv = gst::ElementFactory::make(convert_name, None).unwrap();
+    let sink = gst::ElementFactory::make(sink_name, None).unwrap();
+
+    let pipe_bin = pipe.clone().dynamic_cast::<gst::Bin>().unwrap();
```

Comment:
Instead of dynamic casting here, you could do that already (to a `gst::Pipeline` I guess?) where you connect to the `pad-added` signal. That way you only need the cast once down there instead of in all these functions everwhere.